File: src/stores/websocket.ts
// AIDEV-NOTE: WebSocket state management - tracks connection status and provides connect/disconnect actions
import { create } from "zustand";
import type { ConnectionStatus, ConnectionStateValue } from "@/lib/types";
import { getBackend } from "@/lib/backend";

interface WebSocketState {
  status: ConnectionStatus;
  lastUpdate: Date | null;
  isConnecting: boolean;

  // Actions
  setStatus: (status: ConnectionStatus) => void;
  setClobStatus: (status: ConnectionStateValue) => void;
  setRtdsStatus: (status: ConnectionStateValue) => void;
  setLastUpdate: (date: Date) => void;

  // Commands
  connectToRtds: (markets: string[]) => Promise<void>;
  disconnectFromRtds: () => Promise<void>;
  connectToClob: (tokenIds: string[]) => Promise<void>;
  disconnectFromClob: () => Promise<void>;
  refreshStatus: () => Promise<void>;
}

export const useWebSocketStore = create<WebSocketState>((set) => ({
  status: {
    clob: "disconnected",
    rtds: "disconnected",
  },
  lastUpdate: null,
  isConnecting: false,

  setStatus: (status) => set({ status }),

  setClobStatus: (clob) =>
    set((state) => ({ status: { ...state.status, clob } })),

  setRtdsStatus: (rtds) =>
    set((state) => ({ status: { ...state.status, rtds } })),

  setLastUpdate: (lastUpdate) => set({ lastUpdate }),

  connectToRtds: async (markets) => {
    set({ isConnecting: true });
    try {
      const backend = await getBackend();
      await backend.connectRtds(markets);
    } finally {
      set({ isConnecting: false });
    }
  },

  disconnectFromRtds: async () => {
    const backend = await getBackend();
    await backend.disconnectRtds();
  },

  connectToClob: async (tokenIds) => {
    set({ isConnecting: true });
    try {
      const backend = await getBackend();
      await backend.connectClob(tokenIds);
    } finally {
      set({ isConnecting: false });
    }
  },

  disconnectFromClob: async () => {
    const backend = await getBackend();
    await backend.disconnectClob();
  },

  refreshStatus: async () => {
    const backend = await getBackend();
    const status = await backend.getConnectionStatus();
    set({ status });
  },
}));

File: src-tauri/src/commands/websocket.rs
// AIDEV-NOTE: Tauri commands for WebSocket connection management

use tauri::State;
use serde::Serialize;
use crate::WebSocketState;
use crate::websocket::{ConnectionState, RtdsClient, ClobWebSocket};

/// Response for connection status
#[derive(Debug, Serialize)]
pub struct ConnectionStatusResponse {
    pub rtds: ConnectionState,
    pub clob: ConnectionState,
}

/// Connect to RTDS WebSocket for market activity
#[tauri::command]
pub async fn connect_rtds(
    ws_state: State<'_, WebSocketState>,
    markets: Vec<String>,
) -> Result<(), String> {
    // Take out any existing client and disconnect it (outside await)
    let old_client = {
        let mut guard = ws_state.rtds.write();
        guard.take()
    };

    if let Some(mut client) = old_client {
        client.disconnect();
    }

    // Create and start new connection
    let mut client = RtdsClient::new(ws_state.manager.clone());
    client.connect(markets).await;

    // Store the new client
    {
        let mut guard = ws_state.rtds.write();
        *guard = Some(client);
    }

    Ok(())
}

/// Disconnect from RTDS WebSocket
#[tauri::command]
pub fn disconnect_rtds(ws_state: State<'_, WebSocketState>) -> Result<(), String> {
    let mut rtds_guard = ws_state.rtds.write();

    if let Some(mut client) = rtds_guard.take() {
        client.disconnect();
    }

    Ok(())
}

/// Connect to CLOB WebSocket for order book data
#[tauri::command]
pub async fn connect_clob(
    ws_state: State<'_, WebSocketState>,
    token_ids: Vec<String>,
) -> Result<(), String> {
    // Take out any existing client and disconnect it (outside await)
    let old_client = {
        let mut guard = ws_state.clob.write();
        guard.take()
    };

    if let Some(mut client) = old_client {
        client.disconnect();
    }

    // Create and start new connection
    let mut client = ClobWebSocket::new(ws_state.manager.clone());
    client.connect(token_ids).await;

    // Store the new client
    {
        let mut guard = ws_state.clob.write();
        *guard = Some(client);
    }

    Ok(())
}

/// Disconnect from CLOB WebSocket
#[tauri::command]
pub fn disconnect_clob(ws_state: State<'_, WebSocketState>) -> Result<(), String> {
    let mut clob_guard = ws_state.clob.write();

    if let Some(mut client) = clob_guard.take() {
        client.disconnect();
    }

    Ok(())
}

/// Get current connection status for both WebSockets
#[tauri::command]
pub fn get_connection_status(ws_state: State<'_, WebSocketState>) -> ConnectionStatusResponse {
    ConnectionStatusResponse {
        rtds: ws_state.manager.rtds_state(),
        clob: ws_state.manager.clob_state(),
    }
}
